### PR TITLE
Oron exercise

### PR DESCRIPTION
**Pull request by orontamir**
new feature exercise in calendar

### CODE REVIEW

**Review by PureDreamer — COMMENTED** [review]
very nice work, it seems you thought about it well!

File: app/routers/exercise.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    tags=["exercise"],
+    responses={404: {"description": "Not found"}},
+)
+
+
+@router.get("/")
+async def exercise(
+        request: Request,
+        session=Depends(get_db),
+        new_user=Depends(get_placeholder_user)):
```
[the remark this answers is not in the document]
Comment (orontamir) [reply app/routers/exercise.py:21]: Fix the issue

File: app/routers/exercise.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    If is active exercise = True
+    Show user exercise for specific day
+    """
+    user = session.query(User).filter_by(id=1).first()
+    if not user:
+        # create default user
+        session.add(new_user)
+        session.commit()
+        user = session.query(User).filter_by(id=1).first()
```
[the remark this answers is not in the document]
Comment (orontamir) [reply app/routers/exercise.py:31]: Fix the issue

File: app/routers/profile.py [hunk 3]
```diff
@@ -52,6 +54,38 @@ async def profile(
     })
 
 
+@router.get("/start_exercise")
+async def start_exercise(session=Depends(get_db)):
+    user = session.query(User).filter_by(id=1).first()
+    is_active_exercise = True
+
+    # Update database
+    user.is_active_exercise = is_active_exercise
```
[the remark this answers is not in the document]
Comment (orontamir) [reply app/routers/profile.py]: Fix the issue

File: app/routers/user_exercise.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    else:
+        user_exercise = update_user_exercise(user, session)
+    return user_exercise
+
+
+def does_user_exercise_exist(session: Session, userid: int) -> bool:
+    """
+      Checking if user exercise with user id is exist
+    """
+    return len(get_user_exercise(session=session, user_id=userid)) == 1
```
[the remark this answers is not in the document]
Comment (orontamir) [reply app/routers/user_exercise.py]: Fix the issue

File: app/main.py [hunk 5]
```diff
@@ -46,11 +46,13 @@ def create_tables(engine, psql_environment):
     search.router,
     telegram.router,
     whatsapp.router,
+    exercise.router,
```
Comment (ivarshav) [inline app/main.py:49]: Please keep alphabetic order :)

File: app/routers/exercise.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    prefix="/exercise",
+    tags=["exercise"],
+    responses={404: {"description": "Not found"}},
+)
+
+
+@router.get("/")
+async def exercise(
+        request: Request,
+        session=Depends(get_db)
```
Comment (ivarshav) [inline app/routers/exercise.py:19]: Plese add session type

File: app/routers/exercise.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        ):
+    """
+    If is active exercise = True
```
Comment (ivarshav) [inline app/routers/exercise.py:22]: Consider:
""" Show user exercise for a specific day if `is_active_exercise ` is ture. """

File: app/routers/exercise.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    Show user exercise for specific day
+    """
+    user = session.query(User).filter_by(id=1).first()
+    if not user:
```
Comment (ivarshav) [inline app/routers/exercise.py:26]: I remember I saw creating a default user in other places in code.
Maybe consider extract that to a util or internal function, so all can get or create default user?

File: app/routers/exercise.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        user = User(
+            username='',
+            password='',
+            email=''
+        )
+    # Get user exercise
+    user_exercise = get_user_exercise(session, user_id=user.id)
+    if user_exercise:
+        # Get exercise day
+        date_time_now = datetime.now()
```
Comment (ivarshav) [inline app/routers/exercise.py:37]: maybe use `date.today()`?

File: app/routers/exercise.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        delta = date_time_now - user_exercise[0].start_date
+        # All exercise split to 30 days
```
Comment (ivarshav) [inline app/routers/exercise.py:39]: exercises

File: app/routers/user_exercise.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    if get_user_exercise(session=session, user_id=userid):
+        return True
+    else:
+        return False
+
+
+def get_user_exercise(session: Session, **param) -> list:
+    """Returns user exercise filter by user id."""
+
```
Comment (ivarshav) [inline app/routers/user_exercise.py:35]: Remove blank line in line 35

File: app/routers/user_exercise.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+      Checking if user exercise with user id is exist
+    """
+    if get_user_exercise(session=session, user_id=userid):
+        return True
+    else:
+        return False
+
+
+def get_user_exercise(session: Session, **param) -> list:
+    """Returns user exercise filter by user id."""
```
Comment (ivarshav) [inline app/routers/user_exercise.py:34]: It's actully filter by param, not only user_id.

You can pass user_id exceptly if you want, something like:
def get_user_exercise(session: Session, user_id, **param)

File: app/routers/user_exercise.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+    try:
+        exercise = list(session.query(UserExercise).filter_by(**param))
```
Comment (ivarshav) [inline app/routers/user_exercise.py:37]: It's a list of exercises, so calling the variable exercise**s** makes more sense (same for the function name)

File: app/routers/exercise.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    user = session.query(User).filter_by(id=1).first()
+    if not user:
+        # create empty default user
+        user = User(
+            username='',
+            password='',
+            email=''
+        )
+    # Get user exercise
```
Comment (ivarshav) [inline app/routers/exercise.py:33]: Where do you check `is_active_exercise`  value?

File: app/routers/profile.py [hunk 15]
```diff
@@ -52,6 +54,36 @@ async def profile(
     })
 
 
+@router.get("/start_exercise")
+async def start_exercise(session=Depends(get_db)):
+    user = session.query(User).filter_by(id=1).first()
```
Comment (ivarshav) [inline app/routers/profile.py:59]: Add TODO for the user id

File: app/routers/profile.py [hunk 16]
```diff
@@ -52,6 +54,36 @@ async def profile(
     })
 
 
+@router.get("/start_exercise")
+async def start_exercise(session=Depends(get_db)):
```
Comment (ivarshav) [inline app/routers/profile.py:58]: Please add types

File: app/routers/profile.py [hunk 17]
```diff
@@ -52,6 +54,36 @@ async def profile(
...
+    user = session.query(User).filter_by(id=1).first()
+
+    # Update database
+    user.is_active_exercise = True
+    session.commit()
+
+    # create user exercise
+    print("create user exercise")
```
Comment (ivarshav) [inline app/routers/profile.py:66]: remove print, you can add log if you want

File: app/routers/profile.py [hunk 18]
```diff
@@ -52,6 +54,36 @@ async def profile(
...
+    create_user_exercise(user, session)
+    url = router.url_path_for("profile")
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.get("/stop_exercise")
+async def stop_exercise(session=Depends(get_db)):
+    """
+    Stop exercise
```
Comment (ivarshav) [inline app/routers/profile.py:75]: """Stop exercise - set `is_active_exercise ` to False"""

File: app/routers/profile.py [hunk 19]
```diff
@@ -52,6 +54,36 @@ async def profile(
     })
 
 
+@router.get("/start_exercise")
```
Comment (ivarshav) [inline app/routers/profile.py:57]: start_exercise and stop_exercise have a lot of common code, DRY :)

File: app/routers/user_exercise.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+            )
+        save(user_exercise, session=session)
+    else:
+        user_exercise = update_user_exercise(user, session)
+    return user_exercise
+
+
+def does_user_exercise_exist(session: Session, userid: int) -> bool:
+    """
+      Checking if user exercise with user id is exist
```
Comment (ivarshav) [inline app/routers/user_exercise.py:25]: Consider:  Check if a user exercise for user id exists.

File: app/routers/user_exercise.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+            user_id=user.id,
+            start_date=datetime.now()
+            )
+        save(user_exercise, session=session)
+    else:
+        user_exercise = update_user_exercise(user, session)
+    return user_exercise
+
+
+def does_user_exercise_exist(session: Session, userid: int) -> bool:
```
Comment (ivarshav) [inline app/routers/user_exercise.py:23]: `user_id`

File: tests/test_a_telegram_asyncio.py [hunk 22]
```diff
@@ -6,7 +6,7 @@
 from .asyncio_fixture import today_date
 from .client_fixture import get_test_placeholder_user
 from app.telegram.handlers import MessageHandler, reply_unknown_user
-from app.telegram.keyboards import DATE_FORMAT
+# from app.telegram.keyboards import DATE_FORMAT
```
Comment (ivarshav) [inline tests/test_a_telegram_asyncio.py:9]: ?

File: tests/test_a_telegram_asyncio.py [hunk 23]
```diff
@@ -194,23 +194,23 @@ async def test_no_chosen_day_handler(self):
...
-{chosen_date.strftime('%B %d')}, {chosen_date.strftime('%A')} Events:
-
-From {today_date.strftime('%d/%m %H:%M')} \
-to {(today_date + timedelta(days=2)).strftime('%d/%m %H:%M')}: \
-Cool today event.
-
-From {(chosen_date + timedelta(days=-1)).strftime('%d/%m %H:%M')} \
-to {(chosen_date + timedelta(days=1)).strftime('%d/%m %H:%M')}: \
-Cool (somewhen in two days) event.\n'''
+    """
```
Comment (ivarshav) [inline tests/test_a_telegram_asyncio.py:197]: ?

File: tests/test_a_telegram_asyncio.py [hunk 24]
```diff
@@ -231,16 +231,17 @@ async def test_reply_unknown_user():
 
 
 class TestBotClient:
-
+    """
```
Comment (ivarshav) [inline tests/test_a_telegram_asyncio.py:234]: ?

File: tests/test_user_exercise.py [hunk 25]
```diff
@@ -0,0 +1,37 @@
+from app.routers.user_exercise import create_user_exercise,\
+    does_user_exercise_exist, get_user_exercise
+from app.routers.user import create_user
+
+
+class TestUserExercise:
+
```
Comment (ivarshav) [inline tests/test_user_exercise.py:7]: Remove blank line in line 7

File: tests/test_user_exercise.py [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    def test_create_user_exercise(self, session):
+        _user = create_user(
```
Comment (ivarshav) [inline tests/test_user_exercise.py:9]: Consider using `user` fixture.

File: tests/test_user_exercise.py [hunk 27]
```diff
@@ -0,0 +1,37 @@
+from app.routers.user_exercise import create_user_exercise,\
+    does_user_exercise_exist, get_user_exercise
+from app.routers.user import create_user
+
+
+class TestUserExercise:
+
+    def test_create_user_exercise(self, session):
```
Comment (ivarshav) [inline tests/test_user_exercise.py:8]: Consider change to static method

File: tests/user_fixture.py [hunk 28]
```diff
@@ -18,6 +18,17 @@ def user(session: Session) -> User:
     delete_instance(session, test_user)
 
 
+@pytest.fixture
+def user_exercise(session: Session, user: User) -> UserExercise:
+    test_user_exercise = create_model(
+        session, UserExercise,
+        user_id=11,
```
Comment (ivarshav) [inline tests/user_fixture.py:25]: Why 11 and not user.id?

File: app/routers/exercise.py (hunk 14 above)
Comment (orontamir) [reply app/routers/exercise.py:33]: by default is_active exercise = false
I check it in exercse tab

File: tests/user_fixture.py (hunk 28 above)
Comment (orontamir) [reply tests/user_fixture.py:25]: just for testing

File: app/main.py (hunk 5 above)
Comment (orontamir) [reply app/main.py:49]: fix the issue

File: app/routers/exercise.py (hunk 6 above)
Comment (orontamir) [reply app/routers/exercise.py:19]: fix the issue

File: app/routers/exercise.py (hunk 7 above)
Comment (orontamir) [reply app/routers/exercise.py:22]: fix the issue

File: app/routers/exercise.py (hunk 10 above)
Comment (orontamir) [reply app/routers/exercise.py:39]: fix the issue

File: app/routers/user_exercise.py (hunk 12 above)
Comment (orontamir) [reply app/routers/user_exercise.py:34]: fix the issue

File: app/routers/user_exercise.py (hunk 13 above)
Comment (orontamir) [reply app/routers/user_exercise.py:37]: fix the issue

File: app/routers/profile.py (hunk 16 above)
Comment (orontamir) [reply app/routers/profile.py:58]: fix the issue

File: app/routers/user_exercise.py (hunk 20 above)
Comment (orontamir) [reply app/routers/user_exercise.py:25]: fix the issue

File: app/routers/user_exercise.py (hunk 21 above)
Comment (orontamir) [reply app/routers/user_exercise.py:23]: fix the issue

File: app/routers/user_exercise.py [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+    try:
+        user_exercise = list(session.query(UserExercise).filter_by(**param))
+    except SQLAlchemyError:
+        return []
+    else:
+        return user_exercise
+
+
+def update_user_exercise(user: User, session: Session) -> UserExercise:
+    user_exercise = session.query(UserExercise).filter_by(user_id=user.id)\
```
Comment (yammesicka) [inline app/routers/user_exercise.py:44]: Prefer splitting using parenthesis and not \

File: app/routers/user_exercise.py [hunk 30]
```diff
@@ -0,0 +1,49 @@
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm import Session
+from datetime import datetime
+from app.database.models import User, UserExercise
+from app.internal.utils import save
+
+
+def create_user_exercise(user: User, session: Session) -> UserExercise:
```
Comment (yammesicka) [inline app/routers/user_exercise.py:8]: On some function you put the user first, and on others the session is first. Stick to one convention

**Review by yammesicka — COMMENTED** [review]
Great job! Please take into account the comments of @ivarshav .

Also, Where are the images from? Are there free to use? Should we give credit?

File: app/templates/exercise.html [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+
+{% block content %}
+
+    <div class="container mt-4">
+        <div class="row">
+            <!-- Right side -->
+            <div class="col-5">
+                <!-- Exercise -->
+                {% if user.is_active_exercise %}
+                    <div style="font-family: 'Assistant', sans-serif;text-align: center;">
```
[the remark this answers is not in the document]
Comment (orontamir) [reply app/templates/exercise.html:12]: fix the issue

File: app/routers/user_exercise.py (hunk 29 above)
Comment (orontamir) [reply app/routers/user_exercise.py:44]: fix the issue

File: app/routers/user_exercise.py (hunk 30 above)
Comment (orontamir) [reply app/routers/user_exercise.py:8]: fix the issue

**Comment by orontamir** [comment]
Hi Yam

Thanks for your PR

All images are from the application: "Lose Weight in 30 days" its free to use

**Comment by orontamir** [comment]
Hi Yam

Thanks for your PR

All images are from the application: "Lose Weight in 30 days" its free to use

**Pull request closed without merging**